src/cargo/core/registry.rs

```diff
@@ -335,35 +357,58 @@ http://doc.crates.io/specifying-dependencies.html#overriding-dependencies
 }
 
 impl<'cfg> Registry for PackageRegistry<'cfg> {
-    fn query(&mut self, dep: &Dependency) -> CargoResult<Vec<Summary>> {
+    fn query(&mut self,
+             dep: &Dependency,
+             f: &mut FnMut(Summary)) -> CargoResult<()> {
         // Ensure the requested source_id is loaded
         self.ensure_loaded(dep.source_id(), Kind::Normal).chain_err(|| {
             format!("failed to load source for a dependency \
                      on `{}`", dep.name())
         })?;
 
+        // Look for an override and get ready to query the real source
         let override_summary = self.query_overrides(&dep)?;
-        let real_summaries = match self.sources.get_mut(dep.source_id()) {
-            Some(src) => Some(src.query(&dep)?),
-            None => None,
+        let mut sources = self.sources.borrow_mut();
```

Comment:
>the call to warn_bad_override caught me by surprise.

Hm, shouldn't the borrow of `sources` be dead by the time we call `.warn_bad_override`? If it is the case, then adding a block `{}` could help? Something like this:

<details>

```Rust
        let (override_summary, n, to_warn) = {
            // Look for an override and get ready to query the real source.
            //
            // Note that we're not using `&mut self` to load `self.sources` here but
            // rather we're going through `RefCell::borrow_mut`. This is currently
            // done to have access to both the `lock` function and the
            // `warn_bad_override` functions we call below.
            let override_summary = self.query_overrides(&dep)?;
            let mut sources = self.sources.get_mut();
            let source = match sources.get_mut(dep.source_id()) {
                Some(src) => src,
                None => {
                    if override_summary.is_some() {
                        bail!("override found but no real ones");
                    }
                    return Ok(())
                }
            };

            match override_summary {
                // If we have an override summary then we query the source to sanity
                // check its results. We don't actually use any of the summaries it
                // gives us though.
                Some(s) => {
                    let mut n = 0;
                    let mut to_warn = None;
                    source.query(dep, &mut |summary| {
                        n += 1;
                        to_warn = Some(summary);
                    })?;
                    (s, n, to_warn)
                }

                // If we don't have an override then we just ship everything
                // upstairs after locking the summary
                None => {
                    return source.query(dep, &mut |summary| (unimplemented!()))
                }
            }
        };

        if n > 1 {
            bail!("found an override with a non-locked list");
        } else if let Some(summary) = to_warn {
            self.warn_bad_override(&override_summary, &summary)?;
        }
        f(self.lock(override_summary));
        Ok(())
```

</details>

This variant gets pretty ugly though, so perhaps we can refactor logic here to handle the easy cases up front?

<details>

```Rust
        let (override_summary, n, to_warn) = {
            let overriden = self.query_overrides(&dep)?;
            let real = self.sources.get_mut().get_mut(dep.source_id());
            match (overriden, real) {
                (Some(_), None) => bail!("override found but no real ones"),
                (None, None) => return Ok(()),
                (None, Some(source)) => return source.query(dep, &mut |summary| (unimplemented!())),
                (Some(override_summary), Some(source)) => {
                    let mut n = 0;
                    let mut to_warn = None;
                    source.query(dep, &mut |summary| {
                        n += 1;
                        to_warn = Some(summary);
                    })?;
                    (override_summary, n,  to_warn)
                }
            }
        };

        if n > 1 {
            bail!("found an override with a non-locked list");
        } else if let Some(summary) = to_warn {
            self.warn_bad_override(&override_summary, &summary)?;
        }
        f(self.lock(override_summary));
        Ok(())
```

</details>